### ISSUE TITLE
Issues for NEON: Enabling Efficient Support for Nonlinear Operations in Resistive RAM-based Neural Network Accelerators 

### ISSUE THREAD

**Issue by DingdongD** [adityamanglik/Spider-Gait-Simulator--Genetic-Algorithm#1]
Dear Mr. Aditya Manglik,

I hope this email finds you well. I am reaching out to you regarding some inquiries I have about your paper titled "NEON: Enabling Efficient Support for Nonlinear Operations in Resistive RAM-based Neural Network Accelerators." I am (Your Name), from (Your Institution or Company), currently engaged in research aimed at integrating RRAM-CIM with neural networks to fit exp(x-max(x)), as an alternative to the traditional digital approach of softmax calculation fusion.

In my research, I have encountered some difficulties, particularly in attempting to generalize the fitting of softmax/exp(x-max(x)) for given bounded data, x. I have tried various methods, including fine-tuning the model, freezing parameters, replacing NEON fine-tuning parameters, and using weighted MSE loss to penalize portions of the data with larger fitting errors. However, the improvements have been limited.
![1709105952(1)](https://github.com/adityamanglik/Spider-Gait-Simulator--Genetic-Algorithm/assets/88468868/6bfd3efc-d04e-4620-9095-eaf44ad69f2a)

After reading your paper, I noticed that you achieved precision up to 10e-8 successfully using fine-tuning and MSE loss. I would greatly appreciate the opportunity to inquire about some details of your research, particularly the specific methods and strategies employed to achieve this precision.

If convenient for you, may I ask you some more detailed questions? Your expertise and insights would be invaluable to me, and I am eager to reference your work in mine.

Thank you for taking the time to consider my email. I look forward to your response.

Warmest regards,
Xuliang

**Comment by adityamanglik** [comment]
Dear Mr. DingdongD

Thank you for the question. I am happy that you find my work and paper useful. 
Here are a few questions I used to guide the process of optimization:
1. What dataset are you using to train the neural network? 
2. How many hidden layers does the network have?
3. What is the size of each hidden layer?
4. What is the weight initialization method you used? 

I found the answer to each question to be quite helpful in improving the network's accuracy.
Please let me know if I can be of assistance further.

Best regards
Aditya

**Comment by DingdongD** [comment]
Dear Aditya Manglik,
Thank you for your prompt response. I appreciate the opportunity to address my queries regarding my experiments. Below, I provide detailed responses to each of your questions:
Regarding the data set generation, I experimented with two approaches:
Approach 1: Generating data (X) within the interval [a, b] using either uniform distribution or Gaussian distribution (rand/randn). The labels (Y) were generated through the mapping function exp(x-max(x)).
Approach 2: Utilizing specific task datasets such as CIFAR10, MNIST, etc. In this approach, I embedded standard softmax or MLP-softmaxNet followed by the backbone. Initially, I trained the model's performance on downstream tasks until a certain threshold. Following this, I froze the parameters of the backbone layers and fine-tuned the parameters of the MLP-softmaxNet using MSE loss. However, both approaches yielded minimal effects. The minimum MSE loss achieved in the first approach was in the order of 0.003, while in the second approach, it was in the order of 0.01.

To adapt to the attention layers' hidden size in the transformer, I opted for a hidden layer size of 128/196, matching the input size, with a depth ranging from 1 to 3. Despite attempting to increase the hidden layer's size and depth, it did not yield significant improvements.

As for the initialization methods, I experimented with both Xavier uniform and uniform initialization. However, the results did not show significant improvement.
Thank you very much for considering my request. I eagerly await your response and look forward to the opportunity to discuss these matters further.
Best regards
Xuliang

**Comment by DingdongD** [comment]
``
Method 1:
def generate_softmax_exp_data(batch, dimension):
    # X = torch.randn(batch, dimension)

    a, b = -3.78, 1.2
    X = []
    for i in range(batch):
        x = (b - a) * np.random.rand(dimension) + a
        X.append(torch.from_numpy(x).float())  # raw
        x = np.sort(x)
        X.append(torch.from_numpy(x).float())  # sort
        np.random.shuffle(x)
        X.append(torch.from_numpy(x).float())  # shuffle

    X = torch.stack(X)
    max_values = torch.max(X, dim=1, keepdim=True)[0]
    # X = torch.stack([(b - a) * torch.rand(dimension) + a for _ in range(batch)])
    # print(X.shape)
    # print("X,", X[0])
    # print("max:",max_values)
    out = torch.exp(X - max_values)
    out = out.detach()
    # out = torch.exp(X)
    # out = F.softmax(X, dim=-1)
    return X, out

**Comment by DingdongD** [comment]
``
def get_dataloader(samples, batch, dimension):
    X1, Y1 = generate_softmax_data(samples, dimension)
    # X1, Y1 = generate_single_data(samples, dimension, thres=0.2)
    # train_dataset = TensorDataset(X1, Y1)
    # visualize_data_distribution(Y1)

    train_size = int(0.8 * len(X1))
    val_size = len(X1) - train_size
    train_dataset, test_dataset = random_split(TensorDataset(X1, Y1), [train_size, val_size])

    # X2, Y2 = generate_softmax_data(samples//2, dimension)
    # test_dataset = TensorDataset(X2, Y2)
    train_dataloader = DataLoader(train_dataset, batch_size=batch, shuffle=True)
    test_dataloader = DataLoader(test_dataset, batch_size=batch, shuffle=False)

    return train_dataloader, test_dataloader

**Comment by DingdongD** [comment]
class ExpNet(nn.Module):
    def __init__(self, hidden_size1, hidden_size2, layer_num):
        super().__init__()

        self.layer_num = layer_num
        # self.conv1 = nn.Conv2d(1, 1, 3, 1, 1)
        self.fc1 = nn.Linear(hidden_size1, hidden_size2)
        self.fc2 = nn.ModuleList([nn.Linear(hidden_size2, hidden_size2) for _ in range(layer_num)])
        self.fc3 = nn.Linear(hidden_size2, hidden_size1)

        # nn.init.xavier_uniform_(self.fc1.weight)
        # for layer in self.fc2:
        #     nn.init.xavier_uniform_(layer.weight)
        # nn.init.xavier_uniform_(self.fc3.weight)
        nn.init.kaiming_uniform_(self.fc1.weight, mode='fan_in', nonlinearity='relu')
        for layer in self.fc2:
            nn.init.kaiming_uniform_(layer.weight, mode='fan_in', nonlinearity='relu')
            nn.init.constant_(layer.bias, 0)
        nn.init.kaiming_uniform_(self.fc3.weight, mode='fan_in', nonlinearity='relu')
        nn.init.constant_(self.fc1.bias, 0)
        nn.init.constant_(self.fc3.bias, 0)

        # self.reset_parameters()

    def reset_parameters(self):
        for layer in [self.fc1] + list(self.fc2) + [self.fc3]:
            nn.init.uniform_(layer.weight, -0.1, 0.1)
            nn.init.uniform_(layer.bias)

    def forward(self, x):

        x1 = F.relu(self.fc1(x))
        # x1 = F.leaky_relu(self.fc1(x), negative_slope=0.01)
        for layer in self.fc2:
            x2 = F.relu(layer(x1))
            # x2 = F.leaky_relu(layer(x1), negative_slope=0.01)
        # out = F.relu(self.fc3(x2))
        # out = F.leaky_relu(self.fc3(x2), negative_slope=0.01)
        out = F.relu(self.fc3(x2))
        return out

**Comment by DingdongD** [comment]
#!/user/bin/env python3
# -*- coding: utf-8 -*-


"""
This File is to train quantized ExpNet
"""
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import copy
import logging
import argparse
import os
import time
from os.path import join as pjoin

import random
import numpy as np

from datetime import timedelta

import torch
import torch.distributed as dist
from matplotlib import pyplot as plt
from torchvision.transforms import transforms

from tqdm import tqdm
from torch.utils.tensorboard import SummaryWriter
from torchvision import models
import torch.nn.functional as F
from explib import *

np.set_printoptions(suppress=True, precision=6)


def get_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("--batch_size", default=64, type=int,
                        help="The minimum samples in each training")
    parser.add_argument("--samples", default=2000, type=int,
                        help="The minimum samples in each training")
    parser.add_argument("--dimension", default=768, type=int,
                        help="The minimum samples in each training")
    parser.add_argument("--epochs", default=50, type=int,
                        help="Training Epoch")
    parser.add_argument("--lr", default=1e-3, type=float,
                        help="Learning rate")
    parser.add_argument("--layer_num", default=1, type=int,
                        help="The number of intermediate layers")
    parser.add_argument("--output_dir", default="ExpNet_Out_Log", type=str,
                        help="The output directory"
                             " where checkpoints will be written.")
    parser.add_argument("--mode", default="quant", type=str,
                        help="whether to quant")
    args = parser.parse_args()
    return args


def eval(args, test_dataloader, model, criterion):
    criterion = nn.MSELoss()
    model.eval()
    avg_loss = 0.
    with torch.no_grad():
        for inputs, labels in test_dataloader:
            inputs, labels = inputs.to(args.device), labels.to(args.device)
            outputs = model(inputs)
            loss = criterion(outputs, labels)
            avg_loss += loss.item()
    avg_loss /= len(test_dataloader)

    print("Test dataset loss is:", avg_loss)

    return avg_loss


def train(args, train_dataloader, test_dataloader,
          model, criterion, optimizer):
    num_epochs = args.epochs
    best_loss = 1e6
    model_path = '/home/xuliangyu/Codes_24/AESHA/' + args.output_dir
    if not os.path.exists(model_path):
        os.mkdir(model_path)
    model_path = model_path + '/model.pth'

    model.train()
    for epoch in range(num_epochs):
        for inputs, labels in train_dataloader:
            # inputs = inputs.unsqueeze(1)
            inputs, labels = inputs.to(args.device), labels.to(args.device)

            outputs = model(inputs)
            loss = criterion(outputs, labels)

            optimizer.zero_grad()
            # loss.requires_grad_(True)
            loss.backward()
            # loss.backward(retain_graph=True)
            optimizer.step()

        if (epoch + 1) % 10 == 0:
            print(f'Epoch [{epoch + 1}/{num_epochs}], Train Loss:{loss.item()}')
            avg_loss = eval(args, test_dataloader, model, criterion)

            if best_loss > avg_loss:
                best_loss = avg_loss
                # best_model = model
                print("Updating best model with loss: ", best_loss)
                torch.save(model.state_dict(), model_path)

            model.train()


class CustomLoss(nn.Module):
    def __init__(self):
        super(CustomLoss, self).__init__()
        self.threshold = nn.Parameter(torch.tensor(0.001), requires_grad=True)
        self.large_penalty = nn.Parameter(torch.tensor(2.0), requires_grad=True)
        self.small_penalty = nn.Parameter(torch.tensor(0.1), requires_grad=True)
        self.para = nn.Parameter(torch.tensor(0.5), requires_grad=True)
        self.mse_loss = nn.MSELoss()
        self.l1_loss = nn.L1Loss()

    def forward(self, true_values, predicted_values):
        # residuals = (true_values - predicted_values) ** 2
        #
        # penalty = torch.where(residuals > self.threshold,
        #                       self.large_penalty * residuals,
        #                       self.small_penalty * residuals)
        # # print(residuals)
        # print("penalty:", penalty)
        mse = self.mse_loss(true_values, predicted_values)
        l1 = self.l1_loss(true_values, predicted_values)
        return l1 + mse

class LogCoshLoss(nn.Module):
    def __init__(self):
        super(LogCoshLoss, self).__init__()

    def forward(self, true_values, predicted_values):
        residual = torch.log(predicted_values + 1) - torch.log(true_values + 1)
        loss = residual ** 2
        return torch.sqrt(torch.mean(loss))


def main():
    args = get_args()
    device = torch.device("cuda:{}".format("1"))
    args.device = device
    train_loader, valid_loader = get_dataloader(args.samples,
                                                args.batch_size,
                                                args.dimension)
    hidden_size1 = args.dimension
    hidden_size2 = args.dimension
    layer_num = args.layer_num

    if args.mode != 'quant':
        model = ExpNet(hidden_size1, hidden_size2, layer_num).to(args.device)
    else:
        model = QuantExpNet(hidden_size1, hidden_size2, layer_num).to(args.device)

    # criterion = nn.MSELoss()
    criterion = CustomLoss()
    # criterion = LogCoshLoss()
    optimizer = torch.optim.Adam(model.parameters(), lr=args.lr, weight_decay=1e-4)
    train(args, train_loader, valid_loader,
          model, criterion, optimizer)


def predict():
    args = get_args()
    device = torch.device("cuda:{}".format("1"))
    args.device = device
    train_loader, valid_loader = get_dataloader(args.samples,
                                                args.batch_size,
                                                args.dimension)
    hidden_size1 = args.dimension
    hidden_size2 = args.dimension
    layer_num = args.layer_num

    if args.mode != 'quant':
        model = ExpNet(hidden_size1, hidden_size2, layer_num)
    else:
        model = QuantExpNet(hidden_size1, hidden_size2, layer_num)
    model_path = '/home/xuliangyu/Codes_24/AESHA/' + args.output_dir + '/model.pth'
    model_para = torch.load(model_path)
    model.load_state_dict(model_para)

    # data = torch.randn(1, args.dimension)
    a, b = -3.78, 1.2
    data = (b - a) * torch.rand(1, args.dimension) + a
    max_values = torch.max(data, dim=1, keepdim=True)[0]
    out1 = torch.exp(data - max_values)
    # out1 = F.softmax(data, dim=-1)
    out2 = model(data)

    print(data)
    print(out1)
    print(out2)
    # print(torch.sum(out2, dim=1))

    out2 = out2.detach().numpy()
    print(data[0])
    plt.scatter(data[0], out1[0], label='Actual Function')
    plt.scatter(data[0], out2[0], label='Predicted Function')
    plt.legend()
    plt.show()


if __name__ == '__main__':
    main()
    predict()

    # a = torch.randn(10)
    # mask = torch.randint(0, 2, (10,)).float()
    # b = a - (1 - mask) * 1e6
    # print(b)
    # print(mask)
    # print(torch.mul(mask, a))
    # print(a)
    # out1 = F.softmax(a, dim=0)
    # mask_out = F.softmax(b, dim=0)
    # print("raw attn:\n", out1)
    # print("mask attn:\n", mask_out)
    # re_attn = torch.mul(mask, out1)
    # re_attn = re_attn.div_(torch.sum(re_attn))
    # print("re attn:\n", re_attn)

    # import pickle
    # kset_path = '/home/xuliangyu/Codes_24/AESHA/cifar10_fintune_vitb_0226/best/best_thres.pkl'
    # with open(kset_path, 'rb') as f:
    #     kset = pickle.load(f)
    #     print(kset)
